kedro/ipython/__init__.py

```diff
@@ -21,46 +20,75 @@
 from kedro.framework.startup import _is_project, bootstrap_project
 
 logger = logging.getLogger(__name__)
-default_project_path = Path.cwd()
 
 
-def _remove_cached_modules(package_name):  # pragma: no cover
-    to_remove = [mod for mod in sys.modules if mod.startswith(package_name)]
-    # `del` is used instead of `reload()` because: If the new version of a module does not
-    # define a name that was defined by the old version, the old definition remains.
-    for module in to_remove:
-        del sys.modules[module]
+def load_ipython_extension(ipython):
+    """
+    Main entry point when %load_ext kedro.ipython is executed, either manually or
+    automatically through `kedro ipython` or `kedro jupyter lab/notebook`.
+    IPython will look for this function specifically.
+    See https://ipython.readthedocs.io/en/stable/config/extensions/index.html
+    """
+    ipython.register_magic_function(magic_reload_kedro, magic_name="reload_kedro")
 
+    # Check if a project path can be resolved without input, if not, return early
+    project_path = _resolve_project_path()
+    if project_path is None:
+        logger.warning(
+            "Kedro extension was registered but couldn't find a Kedro project. "
+            "Make sure you run '%reload_kedro <project_root>'."
+        )
+        return
 
-def _find_kedro_project(current_dir: Path):  # pragma: no cover
-    while current_dir != current_dir.parent:
-        if _is_project(current_dir):
-            return current_dir
-        current_dir = current_dir.parent
+    reload_kedro()
 
-    return None
+
+@needs_local_scope
+@magic_arguments()
+@argument(
+    "path",
+    type=str,
+    help=(
+        "Path to the project root directory. If not given, use the previously set"
+        "project root."
+    ),
+    nargs="?",
+    default=None,
+)
+@argument("-e", "--env", type=str, default=None, help=ENV_HELP)
+@argument(
+    "--params",
+    type=lambda value: _split_params(None, None, value),
+    default=None,
+    help=PARAMS_ARG_HELP,
+)
+def magic_reload_kedro(line: str, local_ns: Dict[str, Any] = None):
+    """
+    The `%reload_kedro` IPython line magic. See
+     https://kedro.readthedocs.io/en/stable/tools_integration/ipython.html for more.
+    """
+    args = parse_argstring(magic_reload_kedro, line)
+    reload_kedro(args.path, args.env, args.params, local_ns)
 
 
 def reload_kedro(
-    path: str = None, env: str = None, extra_params: Dict[str, Any] = None
-):  # pragma: no cover
+    path: str = None,
+    env: str = None,
+    extra_params: Dict[str, Any] = None,
+    local_ns: Optional[Dict[str, Any]] = None,
+) -> None:  # pragma: no cover
     """Function that underlies the %reload_kedro Line magic. This should not be imported
     or run directly but instead invoked through %reload_kedro."""
 
     # If a path is provided, set it as default for subsequent calls
```

Comment:
This is not always true now - if a user is using `reload_kedro` instead of `%reload_kedro`. It only remembers the path when it's using the line magic

Reply:
You're right, fixed this in [72509ec](https://github.com/kedro-org/kedro/pull/1955/commits/72509ec930513d8992818df69b25233d1f48bfeb)